plugin/Utils.pm

```diff
@@ -79,69 +83,94 @@ sub getStationList {
     return;
   }
 
-  my $withWebsvc = sub {
-    my ($websvc) = @_;
-    $log->info('fetching station list');
-    $websvc->getStationList(sub {
-      my (%r) = @_;
-      my ($result, $error) = @r{'result', 'error'};
-      if ($result) {
-        my $stationList = {
-          expiresAt => time() + $STATIONLIST_LIFETIME,
-          stations  => $result->{'stations'},
-        };
-        $cache{'stationList'} = $stationList;
-        $successCb->($stationList->{'stations'}, $websvc);
-      }
-      else {
-        if (ref $error eq 'HASH') {
-          if ($error->{'apiCode'} == 1001) {
-            # auth token expired, clear our cached credentials
-            $log->info('auth token expired, clearing cache');
-            %cache = ();
-          }
-          $error = $error->{'message'}
-        }
-        $log->error($error);
-        $errorCb->("Error getting station list ($error)");
-      }
-    },
-      includeStationArtUrl => JSON::true(),
-      returnAllStations => JSON::true(),
-    );
-  };
-
-  my $withoutWebsvc = sub {
-    $log->error(@_);
-    $errorCb->('Unable to connect/login to Pandora');
-  };
-
-  getWebService($withWebsvc, $withoutWebsvc);
+  # Skip fetching QuickMix/Shuffle station if desired
+  if ($prefs->get('disableQuickMix')) {
+    getRestStationList({
+      stations  => \@stations,
+      successCb => $successCb,
+      errorCb   => $errorCb,
+    });
+  }
+  else {
+    getRestQuickMix({
+      stations  => \@stations,
+      successCb => $successCb,
+      errorCb   => $errorCb,
+    });
+  }
 }
 
+sub getRestStationList {
+  _rest(
+    'v1/station/getStations',
+    {
+      pageSize => $STATIONLIST_PAGESIZE,
+    },
+    \&restStationCallback,
+    \&restErrorCallback,
+    @_,
+  );
+}
 
-sub getStationToken {
-  my ($stationId, $successCb, $errorCb) = @_;
-
-  my $withStationList = sub {
-    my ($stationList, $websvc) = @_;
-    my ($station) = grep { $stationId == $_->{stationId} } @$stationList;
-    if ($station) {
-      $successCb->($station->{stationToken}, $websvc);
-    }
-    else {
-      $log->error('stationId not found in station list');
-      $errorCb->('Station not found');
-    }
-  };
+sub restStationCallback {
+  my (%response) = @_;
+  my $stations = $response{'passthrough'}->{'stations'};
+  $log->debug("Retrieved " . scalar(@{$response{'result'}->{'stations'}}) . " stations");
+  $log->debug("Station IDs: " . join(', ', map { $_->{'stationId'} } @{$response{'result'}->{'stations'}}));
+  my $total_stations = $response{'result'}->{'totalStations'};
+  $log->debug("Total stations: ${total_stations}");
+  $log->debug(Dumper($response{'result'}->{'stations'}->[0]));
+  push(@$stations, @{$response{'result'}->{'stations'}});
+  $log->debug("Total retrieved stations: " . scalar(@$stations));
+  if (scalar(@$stations) < $total_stations) {
+    _rest(
+      'v1/station/getStations',
+      {
+        pageSize   => $STATIONLIST_PAGESIZE,
+        startIndex => scalar(@$stations),
```

Comment:
This offset needs to account for the QuickMix station if included. Fix coming soon.